pkg/batcher/batcher.go

```diff
@@ -0,0 +1,199 @@
+func NewBatcher[T input, U output](ctx context.Context, options Options[T, U]) *Batcher[T, U] {
+	b := &Batcher[T, U]{
+		ctx:      ctx,
+		requests: map[uint64][]*request[T, U]{},
+		trigger:  make(chan struct{}),
+		options:  options,
+	}
+	go b.run()
+	return b
+}
+
+// Add will add an input to the batcher using the batcher's hashing function
+func (b *Batcher[T, U]) Add(ctx context.Context, input *T) Result[U] {
+	request := &request[T, U]{
+		ctx:   ctx,
+		hash:  b.options.RequestHasher(ctx, input),
+		input: input,
+		// The requestor channel is buffered to ensure that the exec runner can always write the result out preventing
+		// any single caller from blocking the others. Specifically since we register our request and then trigger, the
+		// request may be processed while the triggering blocks.
+		requestor: make(chan Result[U], 1),
+	}
+	b.mu.Lock()
+	b.requests[request.hash] = append(b.requests[request.hash], request)
+	b.mu.Unlock()
+	b.trigger <- struct{}{}
+	return <-request.requestor
+}
+
+// DefaultHasher will hash the entire input
+func DefaultHasher[T input](ctx context.Context, input *T) uint64 {
+	hash, err := hashstructure.Hash(input, hashstructure.FormatV2, &hashstructure.HashOptions{SlicesAsSets: true})
+	if err != nil {
+		panic("error hashing")
+	}
+	return hash
+}
+
+// OneBucketHasher will return a constant hash and should be used when there is only one type of request
+func OneBucketHasher[T input](ctx context.Context, input *T) uint64 {
+	return 0
+}
+
+func (b *Batcher[T, U]) run() {
+	for {
+		select {
+		// context that we started with has completed so the app is shutting down
+		case <-b.ctx.Done():
+			return
+		case <-b.trigger:
+			// wait to start the batch of create fleet calls
+		}
+		b.waitForIdle()
+		b.runCalls()
+	}
+}
+
+func (b *Batcher[T, U]) waitForIdle() {
+	timeout := time.NewTimer(b.options.MaxTimeout)
+	idle := time.NewTimer(b.options.IdleTimeout)
+	count := 0
+	for {
+		select {
+		case <-b.trigger:
+			count++
+			if b.options.MaxItems != 0 && count >= b.options.MaxItems {
+				return
+			}
+			if !idle.Stop() {
+				<-idle.C
+			}
+			idle.Reset(b.options.IdleTimeout)
+		case <-timeout.C:
+			return
+		case <-idle.C:
+			return
+		}
+	}
+}
+
+func (b *Batcher[T, U]) runCalls() {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	requestMap := b.requests
+	b.requests = map[uint64][]*request[T, U]{}
+
+	for _, requestBatch := range requestMap {
+		firstInput, err := deepCopy(requestBatch[0].input)
+		if err != nil {
+			logging.FromContext(requestBatch[0].ctx).Infof("error copying input, %s", err)
+			// continue without copying so that we don't fail, modification may occur on the calling input
+			firstInput = requestBatch[0].input
+		}
+		inputs := lo.Map(requestBatch, func(req *request[T, U], _ int) *T { return req.input })
+		requestIdx := -1
+		for _, result := range b.options.BatchExecutor(requestBatch[0].ctx, *firstInput, inputs) {
+			requestIdx++
```

Comment:
optional thought: idiomatic for loops increment at the end of the loop. This might allow you to start at 0. Not sure if helpful.